File: renderer/flutter/lib/serialization/serialization_tag.dart
import 'dart:core';

extension CharConvertion on String {
  int toChar() {
    return codeUnitAt(0);
  }
}

class SerializationTag {
  static const int kUnknownTag = -1;

  static final int kVersion = 0xFF;
  static final int kTrue = 'T'.toChar();
  static final int kFalse = 'F'.toChar();
  static final int kUndefined = '_'.toChar();
  static final int kNull = '0'.toChar();
  static final int kInt32 = 'I'.toChar();
  static final int kUInt32 = 'U'.toChar();
  static final int kDouble = 'N'.toChar();
  static final int kBigInt = 'Z'.toChar();
  static final int kUtf8String = 'S'.toChar();
  static final int kOneByteString = '"'.toChar();
  static final int kTwoByteString = 'c'.toChar();
  static final int kPadding = 0;
  static final int kDate = 'D'.toChar();
  static final int kTrueObject = 'y'.toChar();
  static final int kFalseObject = 'x'.toChar();
  static final int kNumberObject = 'n'.toChar();
  static final int kBigIntObject = 'z'.toChar();
  static final int kStringObject = 's'.toChar();
  static final int kRegexp = 'R'.toChar();
  static final int kArrayBuffer = 'B'.toChar();
  static final int kSharedArrayBuffer = 'u'.toChar();
  static final int kArrayBufferTransfer = 't'.toChar();
  static final int kArrayBufferView = 'V'.toChar();
  static final int kBeginJsMap = ';'.toChar();
  static final int kEndJsMap = ':'.toChar();
  static final int kBeginJsSet = '\''.toChar();
  static final int kEndJsSet = ','.toChar();
  static final int kBeginJsObject = 'o'.toChar();
  static final int kEndJsObject = '{'.toChar();
  static final int kBeginSparseJsArray = 'a'.toChar();
  static final int kEndSparseJsArray = '@'.toChar();
  static final int kBeginDenseJsArray = 'A'.toChar();
  static final int kEndDenseJsArray = '\$'.toChar();
  static final int kTheHole = '-'.toChar();
  static final int kObjectReference = '^'.toChar();
  static final int kWasmModuleTransfer = 'w'.toChar();
  static final int kHostObject = '\\'.toChar();
  static final int kWasmMemoryTransfer = 'm'.toChar();
  static final int kError = 'r'.toChar();

  final int tag;

  SerializationTag(this.tag);
}

class ArrayBufferViewTag {
  static const int kUnknownTag = -1;

  static final int kInt8Array = 'b'.toChar();
  static final int kUInt8Array = 'B'.toChar();
  static final int kUInt8ClampedArray = 'C'.toChar();
  static final int kInt16Array = 'w'.toChar();
  static final int kUInt16Array = 'W'.toChar();
  static final int kInt32Array = 'd'.toChar();
  static final int kUInt32Array = 'D'.toChar();
  static final int kFloat32Array = 'f'.toChar();
  static final int kFloat64Array = 'F'.toChar();
  static final int kDataView = '?'.toChar();

  static int fromTag(int tag) {
    if (tag != kInt8Array &&
        tag != kUInt8Array &&
        tag != kUInt8ClampedArray &&
        tag != kInt16Array &&
        tag != kUInt16Array &&
        tag != kInt32Array &&
        tag != kUInt32Array &&
        tag != kFloat32Array &&
        tag != kFloat64Array &&
        tag != kDataView) {
      return kUnknownTag;
    }
    return tag;
  }
}

class ErrorTag {
  static const int kUnknownTag = -1;

  static final int kEvalError = 'E'.toChar();
  static final int kRangeError = 'R'.toChar();
  static final int kReferenceError = 'F'.toChar();
  static final int kSyntaxError = 'S'.toChar();
  static final int kTypeError = 'T'.toChar();
  static final int kURIError = 'U'.toChar();
  static final int kMessage = 'm'.toChar();
  static final int kStack = 's'.toChar();
  static final int kEnd = '.'.toChar();

  static int fromTag(int tag) {
    if (tag != kEvalError &&
        tag != kRangeError &&
        tag != kReferenceError &&
        tag != kSyntaxError &&
        tag != kTypeError &&
        tag != kURIError &&
        tag != kMessage &&
        tag != kStack &&
        tag != kEnd) {
      return kUnknownTag;
    }
    return tag;
  }
}

File: android/sdk/src/main/java/com/tencent/mtt/hippy/serialization/SerializationTag.java
/* Tencent is pleased to support the open source community by making Hippy available.
 * Copyright (C) 2021 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tencent.mtt.hippy.serialization;

/**
 * A tag that determines the type of the serialized value.
 */
public interface SerializationTag {
  byte VOID = 0;
  byte VERSION = (byte) 0xFF;
  byte TRUE = (byte) 'T'; // kTrue
  byte FALSE = (byte) 'F'; // kFalse
  byte UNDEFINED = (byte) '_'; // kUndefined
  byte NULL = (byte) '0'; // kNull
  byte INT32 = (byte) 'I'; // kInt32
  byte UINT32 = (byte) 'U'; // kUint32
  byte DOUBLE = (byte) 'N'; // kDouble
  byte BIG_INT = (byte) 'Z'; // kBigInt
  byte UTF8_STRING = (byte) 'S'; // kUtf8String
  byte ONE_BYTE_STRING = (byte) '"'; // kOneByteString
  byte TWO_BYTE_STRING = (byte) 'c'; // kTwoByteString
  byte PADDING = (byte) '\0'; // kPadding
  byte DATE = (byte) 'D'; // kDate
  byte TRUE_OBJECT = (byte) 'y'; // kTrueObject
  byte FALSE_OBJECT = (byte) 'x'; // kFalseObject
  byte NUMBER_OBJECT = (byte) 'n'; // kNumberObject
  byte BIG_INT_OBJECT = (byte) 'z'; // kBigIntObject
  byte STRING_OBJECT = (byte) 's'; // kStringObject
  byte REGEXP = (byte) 'R'; // kRegExp
  byte ARRAY_BUFFER = (byte) 'B'; // kArrayBuffer
  byte SHARED_ARRAY_BUFFER = (byte) 'u'; // kSharedArrayBuffer
  byte ARRAY_BUFFER_TRANSFER = (byte) 't'; // kArrayBufferTransfer
  byte ARRAY_BUFFER_VIEW = (byte) 'V'; // kArrayBufferView
  byte BEGIN_JS_MAP = (byte) ';'; // kBeginJSMap
  byte END_JS_MAP = (byte) ':'; // kEndJSMap
  byte BEGIN_JS_SET = (byte) '\''; // kBeginJSSet
  byte END_JS_SET = (byte) ','; // kEndJSSet
  byte BEGIN_JS_OBJECT = (byte) 'o'; // kBeginJSObject
  byte END_JS_OBJECT = (byte) '{'; // kEndJSObject
  byte BEGIN_SPARSE_JS_ARRAY = (byte) 'a'; // kBeginSparseJSArray
  byte END_SPARSE_JS_ARRAY = (byte) '@'; // kEndSparseJSArray
  byte BEGIN_DENSE_JS_ARRAY = (byte) 'A'; // kBeginDenseJSArray
  byte END_DENSE_JS_ARRAY = (byte) '$'; // kEndDenseJSArray
  byte SHARED_OBJECT = (byte) 'p'; // kSharedObject
  byte THE_HOLE = (byte) '-'; // kTheHole
  byte OBJECT_REFERENCE = (byte) '^'; // kObjectReference
  byte WASM_MODULE_TRANSFER = (byte) 'w'; // kWasmModuleTransfer
  byte HOST_OBJECT = (byte) '\\'; // kHostObject
  byte WASM_MEMORY_TRANSFER = (byte) 'm'; // kWasmMemoryTransfer
  byte ERROR = (byte) 'r'; // kError
}
